zkevm-circuits/src/evm_circuit/util/math_gadget.rs

```diff
@@ -810,3 +686,154 @@ pub(crate) fn generate_lagrange_base_polynomial<
     }
     numerator * denominator.invert().unwrap()
 }
+
+/// Construct the gadget that checks a * b + c == d (modulo 2**256),
+/// where a, b, c, d are 256-bit words. This can be used by opcode MUL, DIV,
+/// and MOD. For opcode MUL, set c to 0. For opcode DIV and MOD, treat c as
+/// residue and d as dividend.
+///
+/// We execute a multi-limb multiplication as follows:
+/// a and b is divided into 4 64-bit limbs, denoted as a0~a3 and b0~b3
+/// defined t0, t1, t2, t3
+///   t0 = a0 * b0, contribute to 0 ~ 128 bit
+///   t1 = a0 * b1 + a1 * b0, contribute to 64 ~ 193 bit (include the carry)
+///   t2 = a0 * b2 + a2 * b0 + a1 * b1, contribute to above 128 bit
+///   t3 = a0 * b3 + a3 * b0 + a2 * b1 + a1 * b2, contribute to above 192 bit
+///
+/// so t0 ~ t1 include all contributions to the low 256-bit of product, with
+/// a maximum 68-bit radix (the part higher than 256-bit), denoted as carry_hi
+/// Similarly, we define carry_lo as the radix of contributions to the low
+/// 128-bit of the product.
+/// We can slightly relax the constraint of carry_lo/carry_hi to 72-bit and
+/// allocate 9 bytes for them each
+///
+/// Finally we just prove:
+///   t0 + t1 * 2^64 = <low 128 bit of product> + carry_lo
+///   t2 + t3 * 2^64 + carry_lo = <high 128 bit of product> + carry_hi
+///
+/// Last, we sum the parts that are higher than 256-bit in the multiplication
+/// into overflow
+///   overflow = carry_hi + a1 * b3 + a2 * b2 + a3 * b1 + a2 * b3 + a3 * b2
+///              + a3 * b3
+/// In the cases of DIV and MOD, we need to constrain overflow == 0 outside the
+/// MulAddWordsGadget.
+#[derive(Clone, Debug)]
+pub(crate) struct MulAddWordsGadget<F> {
```

Comment:
Can we consider to rename this gadget so that it reflects all of the Math opcodes it can be used for?

Reply:
I think the current gadget name reflects what it does that just multiple two words and add an additional word. So if rename it to sth like `MulDivModGadget`, it will clash with `MulDivModGadget` for opcode gadget

Reply:
That makes sense!